File: serializeanddeserializeinarytree.py
# Definition for a binary tree node.
# class TreeNode(object):
#     def __init__(self, x):
#         self.val = x
#         self.left = None
#         self.right = None

class Codec:

    def serialize(self, root):
        """Encodes a tree to a single string.
        
        :type root: TreeNode
        :rtype: str
        """
        s = []
        def bfs(queue):
            while queue:
                root = queue.pop(0)
                if root:
                    s.append(str(root.val)+",")
                    queue.append(root.left)
                    queue.append(root.right)
                else:
                    s.append('N'+",")
        queue = []
        if root:
            queue.append(root)
        bfs(queue)
        #print "".join(s)
        return "".join(s)
            
    def deserialize(self, data):
        """Decodes your encoded data to tree.
        
        :type data: str
        :rtype: TreeNode
        """
        
        if len(data) == 0:
            return None
        data = data.split(",")
        head = root = TreeNode(int(data[0]))
        q = [root]
        i = 1
        while i!=len(data)-1:
            root = q.pop(0)
            target = data[i]
            if target != 'N':
                root.left = TreeNode(int(data[i]))
                q.append(root.left)
            i += 1
            target = data[i]
            if target != 'N':
                root.right = TreeNode(int(data[i]))
                q.append(root.right)
            i += 1
        return head
        

# Your Codec object will be instantiated and called as such:
# codec = Codec()
# codec.deserialize(codec.serialize(root))

File: serializeanddeserializeinarytree.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode(int x) : val(x), left(NULL), right(NULL) {}
 * };
 */
class Codec {
public:

    // Encodes a tree to a single string.
    string serialize(TreeNode* root) {
        string s = "";
        queue<TreeNode*> que;
        que.push(root);
        while (!que.empty()) {
            int num = que.size();
            for (int i = 0; i < num; i++) {
                auto target = que.front();
                que.pop();
                if (target) {
                    que.push(target->left);
                    que.push(target->right);
                    s += (to_string(target->val) + ",");
                }
                else {
                    s += "#,";
                }
            }
        }
        cout << s << endl;
        return s;
    }

    // Decodes your encoded data to tree.
    TreeNode* deserialize(string data) {
        if (data.length() == 2 and data == "#,") return NULL;
        int pos = data.find(",");
        int val = stoi(data.substr(0, pos));
        data = data.substr(pos+1);
        TreeNode* root = new TreeNode(val);
        queue<TreeNode*> que;
        que.push(root);
        while (que.size()) {
            auto target = que.front();
            que.pop();
            pos = data.find(",");
            if (data.substr(0, pos) != "#") { 
                val = stoi(data.substr(0, pos));
                target->left = new TreeNode(val);
                que.push(target->left);
                
            }
            data = data.substr(pos+1);
            pos = data.find(",");
            if (data.substr(0, pos) != "#") { 
                val = stoi(data.substr(0, pos));
                target->right = new TreeNode(val);
                que.push(target->right);
            }
            data = data.substr(pos+1);
        }
        return root;
    }
};

// Your Codec object will be instantiated and called as such:
// Codec codec;
// codec.deserialize(codec.serialize(root));
